src/backend/distributed/metadata/node_metadata.c

```diff
@@ -428,66 +428,151 @@ citus_disable_node(PG_FUNCTION_ARGS)
 
-		SetNodeState(nodeName, nodePort, isActive);
-		TransactionModifiedNodeMetadata = true;
+	bool onlyConsiderActivePlacements = false;
+	if (NodeGroupHasShardPlacements(workerNode->groupId,
+									onlyConsiderActivePlacements))
+	{
+		ereport(NOTICE, (errmsg(
+							 "Node %s:%d has active shard placements. Some queries "
+							 "may fail after this operation. Use "
+							 "SELECT citus_activate_node('%s', %d) to activate this "
+							 "node back.",
+							 workerNode->workerName, nodePort,
+							 workerNode->workerName,
+							 nodePort)));
 	}
-	PG_CATCH();
+
+	/*
+	 * Delete replicated table placements from the coordinator's metadata,
+	 * but not remotely. That is because one more more of the remote
+	 * nodes might be down. Instead, we let the background worker
+	 * to sync the metadata when possible.
+	 */
+	bool forceRemoteDelete = false;
+	DeleteAllReplicatedTablePlacementsFromNodeGroup(workerNode->groupId,
+													forceRemoteDelete);
+
+	bool localOnly = true;
+	SetAllNodesInGroup(workerNode->groupId, Anum_pg_dist_node_isactive,
+					   BoolGetDatum(isActive), localOnly);
+
+	TransactionModifiedNodeMetadata = true;
+
+	/*
+	 * We have not propagated the metadata changes yet, make sure that all the
+	 * active nodes get the metadata updates. We defer this operation to the
+	 * background worker to make it possible disabling nodes when multiple nodes
+	 * are down.
+	 */
+	if (UnsetMetadataSyncedForAll())
 	{
-		/* CopyErrorData() requires (CurrentMemoryContext != ErrorContext) */
-		MemoryContextSwitchTo(savedContext);
-		ErrorData *edata = CopyErrorData();
+		TriggerMetadataSyncOnCommit();
+	}
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * TriggerSyncMetadataToPrimaryNodes iterates over the active primary nodes,
+ * and triggers the metadata syncs if the node has not the metadata. Later,
+ * maintenance daemon will sync the metadata to nodes.
+ */
+void
+TriggerSyncMetadataToPrimaryNodes(void)
+{
+	List *workerList = ActivePrimaryNonCoordinatorNodeList(ShareLock);
+	bool triggerMetadataSync = false;
 
-		if (ClusterHasKnownMetadataWorkers())
+	WorkerNode *workerNode = NULL;
+	foreach_ptr(workerNode, workerList)
+	{
+		/* if already has metadata, no need to do it again */
+		if (!workerNode->hasMetadata)
 		{
-			ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
-							errmsg("Disabling %s:%d failed", workerNode->workerName,
-								   nodePort),
-							errdetail("%s", edata->message),
-							errhint(
-								"If you are using MX, try stop_metadata_sync_to_node(hostname, port) "
-								"for nodes that are down before disabling them.")));
+			/*
+			 * Let the maintanince deamon do the hard work of syncing the metadata. We prefer
+			 * this because otherwise node activation might fail withing transaction blocks.
+			 */
+			LockRelationOid(DistNodeRelationId(), ExclusiveLock);
+			SetWorkerColumnLocalOnly(workerNode, Anum_pg_dist_node_hasmetadata,
+									 BoolGetDatum(true));
```

Comment:
I'm a bit surprised we're changing the hasmetadata flag here. Why do we do that?

If a node has no metadata we don't really care about syncing. On the other hand, not quite clear what !hasmetadat means I guess.

Reply:
This is actually copy & paste from the other place, see master: https://github.com/citusdata/citus/blob/44204ec9f1c85796950af115002bb2b9c50039e0/src/backend/distributed/commands/function.c#L1096-L1122

This is probably a hack to trigger the bg worker to sync the workers hasmetadata=true flag for this `workerNode`. Instead, we should probably do it inside `SyncMetadataToNodes`.

Probably better to fix separately?

Reply:
That is a very different scenario. We deliberately enabled metadata syncing from create_distributed_function, since that was the only place where we started relying on it. We could probably remove that logic in the future.

The way to trigger metadata sync is to set metadatasynced=false.

Reply:
I think before this hack, we didn't set `hasmetadata=true` on the workers. So, we should be setting locally before we generate the metadata snapshot.

I think the hacky part here is that we should have set this on `SyncMetadataToNodes`, not here?

Reply:
Not sure I follow. TriggerSyncMetadataToPrimaryNodes seems unrelated to setting hasmetadata flag on the workers.

TriggerSyncMetadataToPrimaryNodes is perhaps misnamed. Its primary purpose is to enable metadata syncing for all nodes and it also triggers a metadata sync via TriggerMetadataSyncOnCommit.

We should not enable metadata syncing in citus_disable_node if it was not already enabled.

Reply:
That is already prevented via `UnsetMetadataSyncedForAll()`:
```C
	if (UnsetMetadataSyncedForAll())
	{
		TriggerMetadataSyncOnCommit();
	}
```

Where it returns false if we don't have any metadata synced nodes

Reply:
(followed similar approach with `citus_update_node`)

Reply:
ok, I missed that we do not call TriggerSyncMetadataToPrimaryNodes from citus_disable_node. Got confused by the move.